[cmake][common] Adding optional support for --zcl argument in chip_zapgen() cmake function

dinabenamar opened the pull request:
#### Summary

The aim of this PR is to introduce optional support for passing "--zcl" argument to the zapgen "generate.py" script, without impacting existing example behavior.

It supports three levels of configuration: 
1- **Default behavior** : No "--zcl" argument is passed, preserving current behavior for all examples.
2- **Global opt-in** : If the "CHIP_ENABLE_ZCL_ARG" cmake variable is set to ON, the default ZCL path "${CHIP_ROOT}/src/app/zap-templates/zcl/zcl.json" is used.
3- **Custom override** : Examples can pass a specific "ZCL_PATH" via the "chip_configure_data_model()" API to use a custom "zcl.json".

Example of use to add a custom ZCL_PATH : 

> chip_configure_data_model(app
>         INCLUDE_SERVER
>         ZAP_FILE ${NXP_EXAMPLE_ZAP_DIR}/thermostat_matter_wifi.zap
>         **ZCL_PATH** ${CHIP_ROOT}/src/app/zap-templates/zcl/zcl-with-test-extensions.json
> )

#### Context

This change helps address build issues in freestanding examples, where the "generate.py" (called in chip_zapgen) relies on relative paths embedded in the ".zap" files to locate the "zcl.json" file. These paths may be invalid if the ".zap" file is moved or used outside its original context. Explicitly passing the ZCL path ensures consistent and portable builds.

#### Testing

These changes were tested by building the following configurations : 
- Build example with default behavior, without ZCL_PATH and without CHIP_ENABLE_ZCL_ARG defined. No change required.
- Build example with "CHIP_ENABLE_ZCL_ARG" set to ON. Change required : https://github.com/project-chip/connectedhomeip/pull/40537/files#diff-6fad594babff0528f5e4695f20317f72c20a61f1787baae23fb6d6dfa93fceacR31 
       - Tests in CI : All NXP examples (except all-clusters-app) are using this configuration.
- Build example using a custom zcl file, by providing the path with ZCL_PATH argument via chip_configure_data_model. Example of change required : https://github.com/project-chip/connectedhomeip/pull/40537/files#diff-3996d06f182576c49a1bbda1ecc53d6cdcd8bd613e9c116f9c0abeff534b4693R108 
      - Tests in CI : NXP all-clusters-app example is using this configuration.

Example of command used to build:
- west build -d build_matter -b rdrw612bga examples/thermostat/nxp
- west build -d build_matter -b rdrw612bga examples/all-clusters-app/nxp

dinabenamar committed:
[cmake] Add optional support for --zcl argument in chip_zapgen CMake function

Signed-off-by: Dina Benamar <dina.benamarelmaaroufi@nxp.com>

dinabenamar committed:
[NXP][cmake][common] Set "CHIP_ENABLE_ZCL_ARG" to ON for all NXP examples to allow default ZCL path to be passed as --zcl argument with chip_zapgen

Signed-off-by: Dina Benamar <dina.benamarelmaaroufi@nxp.com>

dinabenamar commented on build/chip/chip_codegen.cmake:
```diff
@@ -180,6 +180,28 @@ function(chip_zapgen TARGET_NAME)
+        # Add --zcl only if a valid path is set.
+        if(ARG_ZCL_PATH)
+            list(APPEND ZAPGEN_ARGS "--zcl" "${ARG_ZCL_PATH}")
+        endif()
```
Update done.

dinabenamar committed:
[cmake] Simplify logic for handling ZCL_PATH

Signed-off-by: Dina Benamar <dina.benamarelmaaroufi@nxp.com>

andy31415 reviewed (changes requested):
Please add documentation for the added arguments given that documentation already exists.

For the testing section in the PR summary, you included a command that invokes a build script that was not modified by this PR. Could you explain how that works? are there local changes? It was not immediately obvious to me how the changes in this PR translate in the command line you gave working differently ... some explanation/links on how things work would help

dinabenamar commented on build/chip/chip_codegen.cmake:
```diff
@@ -139,7 +139,7 @@ endfunction()
     cmake_parse_arguments(ARG
          ""
-         "INPUT;GENERATOR;OUTPUT_PATH;OUTPUT_FILES"
+         "INPUT;GENERATOR;OUTPUT_PATH;OUTPUT_FILES;ZCL_PATH"
```
Documentation update done

dinabenamar commented:
> Please add documentation for the added arguments given that documentation already exists.
> 
> For the testing section in the PR summary, you included a command that invokes a build script that was not modified by this PR. Could you explain how that works? are there local changes? It was not immediately obvious to me how the changes in this PR translate in the command line you gave working differently ... some explanation/links on how things work would help

Thank you for the review. I have updated the documentation accordingly.

As for the testing section, I added in the description the changes required for testing each configuration. These changes are applied on application cmake level, and do not require the command line to be changed.

dinabenamar committed:
[cmake] Updating documentation to explain the usage of ZCL_PATH in chip_zapgen() and chip_configure_data_model()

Signed-off-by: Dina Benamar <dina.benamarelmaaroufi@nxp.com>

dinabenamar committed:
[NXP][cmake] Updating all-clusters-app example to use custom ZCL_PATH argument with chip_configure_data_model

Signed-off-by: Dina Benamar <dina.benamarelmaaroufi@nxp.com>

andy31415 commented on build/chip/chip_codegen.cmake:
```diff
@@ -135,11 +136,13 @@ endfunction()
 #
 #   OUTPUT_FILES - [OUT] output variable will contain the path of generated files.
 #                  suitable to be added within a build target
+# ZCL_PATH      - [OPTIONAL] path to a custom ZCL JSON file. If provided, it overrides the default ZCL path used by generate.py.
```
please make this to be in line with previous docs: alignment of both the argument and then the text is indented on new lines

dinabenamar committed:
[cmake][common] Updating documentation of chip_zapgen and chip_configure_data_model to explain the purpose of using ZCL_PATH argument

Signed-off-by: Dina Benamar <dina.benamarelmaaroufi@nxp.com>
